opnode/l2/driver_state.go

```diff
@@ -0,0 +1,130 @@
+}
+
+type EngineDriverState struct {
+	// Locks the L1 and L2 head changes, to keep a consistent view of the engine
+	headLock sync.RWMutex
+
+	// l1Head tracks the L1 block corresponding to the l2Head
+	l1Head eth.BlockID
+
+	// l2Head tracks the head-block of the engine
+	l2Head eth.BlockID
+
+	// l2Finalized tracks the block the engine can safely regard as irreversible
+	// (a week for disputes, or maybe shorter if we see L1 finalize and take the derived L2 chain up till there)
+	l2Finalized eth.BlockID
+
+	// The L1 block we are syncing towards, may be ahead of l1Head
+	l1Target eth.BlockID
+
+	// Genesis starting point
+	Genesis Genesis
+}
+
+// L1Head returns the block-id (hash and number) of the last L1 block that was derived into the L2 block
+func (e *EngineDriverState) L1Head() eth.BlockID {
+	e.headLock.RLock()
+	defer e.headLock.RUnlock()
+	return e.l1Head
+}
+
+// L2Head returns the block-id (hash and number) of the L2 chain head
+func (e *EngineDriverState) L2Head() eth.BlockID {
+	e.headLock.RLock()
+	defer e.headLock.RUnlock()
+	return e.l2Head
+}
+
+func (e *EngineDriverState) Head() (l1Head eth.BlockID, l2Head eth.BlockID) {
+	e.headLock.RLock()
+	defer e.headLock.RUnlock()
+	return e.l1Head, e.l2Head
+}
+
+func (e *EngineDriverState) UpdateHead(l1Head eth.BlockID, l2Head eth.BlockID) {
+	e.headLock.Lock()
+	defer e.headLock.Unlock()
+	e.l1Head = l1Head
+	e.l2Head = l2Head
+}
+
+func (e *EngineDriverState) RequestUpdate(ctx context.Context, log log.Logger, driver Driver) (l2Updated bool) {
+	refL1, refL2, err := driver.requestEngineHead(ctx)
+	if err != nil {
+		log.Error("failed to request engine head", "err", err)
+		return false
+	}
+	e.headLock.Lock()
+	defer e.headLock.Unlock()
+
+	e.l1Head = refL1
+	e.l2Head = refL2
+	return e.l1Head != refL1 || e.l2Head != refL2
+}
+
+func (e *EngineDriverState) RequestSync(ctx context.Context, log log.Logger, driver Driver) (l2Updated bool) {
+	if e.l1Head == e.l1Target {
+		log.Debug("Engine is fully synced", "l1_head", e.l1Head, "l2_head", e.l2Head)
+		// TODO: even though we are fully synced, it may be worth attempting anyway,
+		// in case the e.l1Head is not updating (failed/broken L1 head subscription)
+		return false
+	}
+	log.Debug("finding next sync step, engine syncing", "l2", e.l2Head, "l1", e.l1Head)
+	nextRefL1, refL2, err := driver.findSyncStart(ctx)
+	if err != nil {
+		log.Error("Failed to find sync starting point", "err", err)
+		return false
+	}
+	if nextRefL1 == e.l1Head {
+		log.Debug("Engine is already synced, aborting sync", "l1_head", e.l1Head, "l2_head", e.l2Head)
+		return false
+	}
+	if l2ID, err := driver.driverStep(ctx, nextRefL1, refL2, e.l2Finalized); err != nil {
+		log.Error("Failed to sync L2 chain with new L1 block", "l1", nextRefL1, "onto_l2", refL2, "err", err)
+		return false
+	} else {
+		e.UpdateHead(nextRefL1, l2ID) // l2ID is derived from the nextRefL1
+	}
+	return e.l1Head != e.l1Target
```

Comment:
I'm a little worried that this doesn't catch up in the case of a sync. The flow as I understand it is

L1 & L2 head on block 10, say L1 Target (or tip) is a block 20
- New L1 Block -> NotifyL1Head fails b/c out of sync, returns false
- Sync ticker fires -> RequestSync advances by one, but still returns false
- Sync Ticker remains cold

Part of it is that I wonder if this sync function should try to advance to head. I think it may be slightly preferred to setting the `syncquickly` correctly b/c it avoids repeated calls to `syncStart` (only changes in `NotifyL1Head`), though maybe we want an interrupt into this function when we get a new L1 head.

Note: I think a very similar bug is present in the pre-refactor form because `syncbackoff` is always called during the sync ticker.

Reply:
> New L1 Block -> NotifyL1Head fails b/c out of sync, returns false

But it does update the `l1Target`

> Sync ticker fires -> RequestSync advances by one, but still returns false

It returns false when the `l1Head == l1Target`: when it's already synced. If it advances by 1, but not synced afterwards, then it's an "update", and it will continue with `syncQuickly()` (faster sync) and `onL2Update` (slow down polling L2 because we are actively changing it).

It's the `!=` I think that's confusing, it returns true in your example, not false. I'll add a comment to clarify.

Always calling `syncBackoff` is fine, it will reset with `syncQuickly` if there's actually something to sync.